### PR TITLE
add BuildName helper func

### PR DESCRIPTION
**Pull request by vtolstov**
Signed-off-by: Vasiliy Tolstov <v.tolstov@unistack.org>

### CODE REVIEW

File: metrics_test.go [hunk 1]
```diff
@@ -7,6 +7,10 @@ import (
 	"time"
 )
 
+func TestBuildName(t *testing.T) {
+	expectNoPanic(t, `my_metric{firstlabel="value2",zerolabel="value3"}`, func() { BuildName("my_metric", "zerolabel", "value3", "firstlabel", "value2") })
```
Comment (cristaloleg) [inline metrics_test.go:11]:
```
expectNoPanic(t, `my_metric{firstlabel="value2",zerolabel="value3"}`, func() { 
	BuildName("my_metric", "zerolabel", "value3", "firstlabel", "value2")
})
```

is easier to read, wdyt?

File: metrics.go [hunk 2]
```diff
@@ -110,3 +113,37 @@ func WriteFDMetrics(w io.Writer) {
...
+func BuildName(name string, labels ...string) string {
+	var b strings.Builder
+
+	if len(labels)%2 != 0 {
+		panic(fmt.Errorf("BUG: invalid labels count: %q", labels))
+	}
+
+	sort.Sort(byKey(labels))
+
+	_, _ = b.WriteString(name)
```
Comment (cristaloleg) [inline metrics.go:135]: `var b strings.Builder` can be here

File: metrics.go [hunk 3]
```diff
@@ -110,3 +113,37 @@ func WriteFDMetrics(w io.Writer) {
...
+func (k byKey) Swap(i, j int) {
+	k[i*2], k[j*2] = k[j*2], k[i*2]
+	k[i*2+1], k[j*2+1] = k[j*2+1], k[i*2+1]
+}
+
+func BuildName(name string, labels ...string) string {
+	var b strings.Builder
+
+	if len(labels)%2 != 0 {
+		panic(fmt.Errorf("BUG: invalid labels count: %q", labels))
```
Comment (cristaloleg) [inline metrics.go:130]: For example `strings.NewReplacer` explains what exactly is not ok with a param.

```
func NewReplacer(oldnew ...string) *Replacer {
	if len(oldnew)%2 == 1 {
		panic("strings.NewReplacer: odd argument count")
	}
	return &Replacer{oldnew: append([]string(nil), oldnew...)}
}
```

File: metrics.go [hunk 4]
```diff
@@ -110,3 +113,37 @@ func WriteFDMetrics(w io.Writer) {
 func UnregisterMetric(name string) bool {
 	return defaultSet.UnregisterMetric(name)
 }
+
+type byKey []string
```
Comment (cristaloleg) [inline metrics.go:117]: Not sure but there can be a comment about sorting behaviour, wdyt?

File: metrics.go [hunk 5]
```diff
@@ -110,3 +113,37 @@ func WriteFDMetrics(w io.Writer) {
...
+		}
+		_, _ = b.WriteString(labels[idx])
+		_, _ = b.WriteString(`="`)
+		_, _ = b.WriteString(labels[idx+1])
+		_, _ = b.WriteString(`"`)
+	}
+	_, _ = b.WriteRune('}')
+
+	return b.String()
+}
```
Comment (cristaloleg) [inline metrics.go:149]: Should we also validate `labels` entries for allowed characters?

If so - throwing panic can be changes to `return nil, fmt.Errorf("BUG: invalid labels count: %q", labels)`

wdyt?

Comment (cristaloleg) [reply metrics.go:149]: And as a helper wi might add `MustBuildName(...) string` which panics.

Comment (vtolstov) [reply metrics.go:153]: don't think that we need MustBuildName as all set.go functions panics on invalid name. So i'm reuse some logic =)

**Review by cristaloleg — APPROVED** [review]

**Comment by vtolstov** [comment]
@valyala @tenmozes can you merge this if it's ok now?

**Comment by vtolstov** [comment]
gentle ping..

File: metrics.go [hunk 6]
```diff
@@ -110,3 +113,42 @@ func WriteFDMetrics(w io.Writer) {
...
+	var b strings.Builder
+	_, _ = b.WriteString(name)
+	_, _ = b.WriteRune('{')
+	for idx := 0; idx < len(labels); idx += 2 {
+		if idx > 0 {
+			_, _ = b.WriteRune(',')
+		}
+		_, _ = b.WriteString(labels[idx])
+		_, _ = b.WriteString(`="`)
+		_, _ = b.WriteString(labels[idx+1])
```
Comment (valyala) [inline metrics.go:144]: strconv.AppendQuote must be used here in order to properly escape double quotes in label values

**Review by valyala — COMMENTED** [review]
Please see https://github.com/VictoriaMetrics/metrics/issues/25#issuecomment-883920289

**Comment by hagen1778** [comment]
Hi @vtolstov! What would be the typical use case for the new helper? 
If it is:
```
metrics.NewSummary(metrics.BuildName("my_metric", "zerolabel", "value3", ",firstlabel", "value2"))
```
then metric name validation happens twice.

If it is:
```
myMetricName := metrics.BuildName("my_metric", "zerolabel", "value3", ",firstlabel", "value2")
myMetric := metrics.NewSummary(myMetricName)
```
then it is unclear why `metrics.BuildName` panics instead of returning an error for processing.

IMHO, in the role of helper I'd prefer smthng like the following:
```
metrics.NewSummaryWithLabels("my_metric", "zerolabel", "value3", ",firstlabel", "value2")
```
or 
```
metrics.NewSummary("my_metric").WithLabels("zerolabel", "value3", ",firstlabel", "value2")
```
wdyt?

**Comment by vtolstov** [comment]
@hagen1778 may be you case is better, but for now i'm move helper to own framework =)

**Pull request closed without merging**